### PR TITLE
feat: add cached router for remote engine client

### PR DESCRIPTION
**Pull request by Rachelint**
# Which issue does this PR close?

Closes #
Part of #492 

# Rationale for this change
 Now, we do route work in each request, that make performance too bad.
So, this pr add a cached router in remote engine client. 

# What changes are included in this PR?
Add cached router for remote engine client.

# Are there any user-facing changes?
None.

# How does this change test
Test manually.

### CODE REVIEW

File: remote_engine_client/src/cached_router.rs [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+            // If not found, do real route and put it into cache, and return then.
+            let mut cache = self.cache.lock().await;
+
+            // Double check here, it may have been put by others.
+            let channel_opt = cache.get(table_ident).cloned();
+            if let Some(channel) = channel_opt {
+                channel
+            } else {
+                self.do_route(table_ident).await?
+            }
```
Comment (ShiKaiWi) [inline remote_engine_client/src/cached_router.rs:61]: I guess it is not a good way to process heavy stuff (`do_route`) with the cache lock held. Some duplicate routing requests are tolerant, so just do route without holding the cache lock.

File: remote_engine_client/src/cached_router.rs [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+        } else {
+            // If not found, do real route and put it into cache, and return then.
+            let mut cache = self.cache.lock().await;
+
+            // Double check here, it may have been put by others.
+            let channel_opt = cache.get(table_ident).cloned();
+            if let Some(channel) = channel_opt {
+                channel
+            } else {
+                self.do_route(table_ident).await?
```
Comment (ShiKaiWi) [inline remote_engine_client/src/cached_router.rs:60]: It seems the new channel has not been updated into the cache?

Comment (Rachelint) [reply remote_engine_client/src/cached_router.rs]: Fix.

File: remote_engine_client/src/cached_router.rs (hunk 1 above)
Comment (Rachelint) [reply remote_engine_client/src/cached_router.rs]: Yes, `duplicate io` or `io with lock` ...
Mainly, I don't know which is more tolerant.
In fact, I think a mechanism similar to `group commit` may be the good way to solve it.

**Review by ShiKaiWi — APPROVED** [review]
LGTM

File: remote_engine_client/src/cached_router.rs [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+            .fail();
+        }
+
+        // Get channel from pool.
+        let endpoint = route_infos
+            .first()
+            .unwrap()
+            .endpoint
+            .clone()
+            .context(RouteNoCause {
```
Comment (chunshao90) [inline remote_engine_client/src/cached_router.rs:105]:
```suggestion
            .with_context(||RouteNoCause {
```

File: remote_engine_client/src/client.rs [hunk 4]
```diff
@@ -57,24 +50,34 @@ impl Client {
             .await
             .context(Rpc {
```
Comment (chunshao90) [inline remote_engine_client/src/client.rs:51]:
```suggestion
            .with_context(||Rpc {
```

File: remote_engine_client/src/client.rs [hunk 5]
```diff
@@ -87,10 +90,21 @@ impl Client {
...
```
Comment (chunshao90) [inline remote_engine_client/src/client.rs:91]:
```suggestion
            .with_context(||Rpc {
```

**Review by chunshao90 — APPROVED** [review]
LGTM

**Pull request merged**